signature/jws/plugin.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/notaryproject/notation-go/spec/signature"
+)
+
+// PluginSigner signs artifacts and generates JWS signatures
+// by delegating the one or both operations to the named plugin,
+// as defined in
+// https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md#signing-interfaces.
+type PluginSigner struct {
+	Runner       plugin.Runner
+	KeyID        string
+	PluginConfig map[string]string
+}
+
+// Sign signs the artifact described by its descriptor, and returns the signature.
+func (s *PluginSigner) Sign(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	metadata, err := s.getMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if metadata.HasCapability(plugin.CapabilitySignatureGenerator) {
+		return s.generateSignature(ctx, desc, opts)
+	} else if metadata.HasCapability(plugin.CapabilityEnvelopeGenerator) {
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *PluginSigner) getMetadata(ctx context.Context) (*plugin.Metadata, error) {
+	out, err := s.Runner.Run(ctx, plugin.CommandGetMetadata, nil)
+	if err != nil {
+		return nil, fmt.Errorf("metadata command failed: %w", err)
+	}
+	metadata, ok := out.(*plugin.Metadata)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect get-plugin-metadata response type '%T'", out)
+	}
+	if err := metadata.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid plugin metadata: %w", err)
+	}
+	return metadata, nil
+}
+
+func (s *PluginSigner) describeKey(ctx context.Context, config map[string]string) (*plugin.DescribeKeyResponse, error) {
+	req := &plugin.DescribeKeyRequest{
+		ContractVersion: plugin.ContractVersion,
+		KeyID:           s.KeyID,
+		PluginConfig:    config,
+	}
+	out, err := s.Runner.Run(ctx, plugin.CommandDescribeKey, req)
+	if err != nil {
+		return nil, fmt.Errorf("describe-key command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.DescribeKeyResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect describe-key response type '%T'", out)
+	}
+	return resp, nil
+}
+
+func (s *PluginSigner) generateSignature(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != key.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Get algorithm associated to key.
+	alg := key.KeySpec.SignatureAlgorithm()
+	if alg == "" {
+		return nil, fmt.Errorf("keySpec %q not supported: ", key.KeySpec)
+	}
+
+	// Generate payload to be signed.
+	payload := packPayload(desc, opts)
+	if err := payload.Valid(); err != nil {
+		return nil, err
+	}
+
+	// Generate signing string.
+	token := jwtToken(alg.JWS(), payload)
+	signing, err := token.SigningString()
```

Comment:
nit, rename `signing` to `payloadToSign` makes it easy to understand, spec uses the same terminology.